libbeat/autodiscover/providers/kubernetes/pod.go

```diff
@@ -146,6 +146,18 @@ func (p *pod) OnUpdate(obj interface{}) {
 		return
 	}
 
+	// here handle the case when a Pod is in `Terminating` phase.
+	// In this case the pod is neither `PodSucceeded` nor `PodFailed` and
+	// hence a stop/start sequence will take place. However we need to
+	// handle this as just a stop event with a cleanup timeout so as to
+	// to give time to any leftover log lines to be collected while the pod is
+	// in `Terminating` phase.
+	if pod.GetObjectMeta().GetDeletionTimestamp() != nil {
+		p.logger.Debugf("Watcher Pod update (being terminated): %+v", obj)
+		time.AfterFunc(p.config.CleanupTimeout, func() { p.emit(pod, "stop") })
```

Comment:
In case the pod has a grace period (`terminationGracePeriodSeconds`) greater than the cleanup timeout, I think that with this code we would be stopping the pod before their containers finish, so we can miss lines logged during termination.
On https://github.com/elastic/beats/pull/14259 I tried to handle this case by [ignoring update events during termination](https://github.com/elastic/beats/pull/14259/files#diff-3b0c1598273f3d8e5f8fc4c5f134ad17R154), instead of directly triggering an stop.
This code was removed on #17223 when we started relying on the status phase. It probably makes sense to rely on both things (phase and deletion timestamp), but we should be sure of stopping harvesters only after the containers have been actually stopped.

Users shouldn't need to increase cleanup timeout to handle pods with high grace periods, cleanup timeout is intended to avoid missing lines logged after the pod has been actually terminated, if filebeat hasn't read them yet.

Reply:
👍 sounds reasonable, thank you for commenting @jsoriano ! I will add this change.